merlin/models/tf/models/base.py

```diff
@@ -812,60 +812,172 @@ def call_train_test(
+            if dense_target_mask is not None:
+                target._keras_mask = dense_target_mask
+
+        if getattr(target, "_keras_mask", None) is not None:
+            # Copies the mask from the targets to the predictions
+            # because Keras considers the prediction mask in loss
+            # and metrics computation
+            prediction._keras_mask = target._keras_mask
+
+        # Ensuring targets and preds have the same dtype
+        target = tf.cast(target, prediction.dtype)
+
+        # If targets are scalars (1-D) and predictions are sequential (3-D),
+        # extract predictions at target position because Keras expects
+        # predictions and targets to have the same shape.
+        if getattr(prediction, "_keras_mask", None) is not None:
+            rank_check = tf.logical_and(
+                tf.equal(tf.rank(target), 1),
+                tf.equal(tf.rank(prediction), 3),
+            )
+            prediction = tf.cond(
+                rank_check, lambda: self._extract_masked_predictions(prediction), lambda: prediction
+            )
+
+        # Ensuring targets are one-hot encoded if they are not
+        target = tf.cond(
+            tf.rank(target) == tf.rank(prediction) - 1,
+            lambda: tf.one_hot(
+                tf.cast(target, tf.int32),
+                tf.shape(prediction)[-1],
+                dtype=prediction.dtype,
+            ),
+            lambda: tf.reshape(target, tf.shape(prediction)),
+        )
+
+        return prediction, target
+
+    def _adjust_ragged_predictions_and_targets(
+        self,
+        prediction: TensorLike,
+        target: tf.Tensor,
+    ):
+        """Adjusts the predictions and targets to ensure compatibility with most Keras losses and metrics.
+        This methods applies the following transformations to the target and prediction tensors:
+        - Select ragged targets based on the mask information, if defined.
+        - Remove mask information from the ragged targets and predictions.
+        - One-hot encode targets if their rank is one less than the rank of predictions.
+        - Ensure that targets have the same shape and dtype as predictions.
+        """
+        if isinstance(target, tf.RaggedTensor) and getattr(target, "_keras_mask", None) is not None:
+            # Select targets at masked positions and return
+            # a ragged tensor.
+            target = tf.ragged.boolean_mask(
+                target.with_row_splits_dtype(target.dtype), target._keras_mask
+            )
+
+        # Ensuring targets and preds have the same dtype
+        target = tf.cast(target, prediction.dtype)
+
+        # If targets are scalars (1-D) and predictions are sequential (3-D),
+        # extract predictions at target position because Keras expects
+        # predictions and targets to have the same shape.
+        if getattr(prediction, "_keras_mask", None) is not None:
+            rank_check = tf.logical_and(
```

Comment:
Do we still need this logic, as Transformer blocks will have the expected blocks to mask the last item and make them 2D?
I mean, couldn't we have the logic inside `_extract_masked_predictions` within `TransformerInferenceHiddenState`, that would work for both inference and evaluation when pre=SequenceMaskLast is used?